snowflake/src/main/java/org/apache/iceberg/snowflake/JdbcSnowflakeClient.java

```diff
@@ -0,0 +1,386 @@
+    List<SnowflakeIdentifier> databases;
+    try {
+      databases =
+          connectionPool.run(
+              conn ->
+                  queryHarness.query(
+                      conn, finalQuery, DATABASE_RESULT_SET_HANDLER, (String[]) null));
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(e, "Failed to check if database '%s' exists", database);
+    } catch (InterruptedException e) {
+      throw new UncheckedInterruptedException(
+          e, "Interrupted while checking if database '%s' exists", database);
+    }
+
+    // Filter to handle the edge case of '_' appearing as a wildcard that can't be remapped the way
+    // it can for predicates in SELECT statements.
+    databases.removeIf(db -> !database.databaseName().equalsIgnoreCase(db.databaseName()));
+    return !databases.isEmpty();
+  }
+
+  @Override
+  public boolean schemaExists(SnowflakeIdentifier schema) {
+    Preconditions.checkArgument(
+        schema.type() == SnowflakeIdentifier.Type.SCHEMA,
+        "schemaExists requires a SCHEMA identifier, got '%s'",
+        schema);
+
+    if (!databaseExists(SnowflakeIdentifier.ofDatabase(schema.databaseName()))) {
+      return false;
+    }
+
+    // Due to current limitations in PreparedStatement parameters for the LIKE clause in
+    // SHOW SCHEMAS queries, we'll use a fairly limited allowlist for identifier characters,
+    // using wildcards for non-allowed characters, and post-filter for matching.
+    final String finalQuery =
+        String.format(
+            "SHOW SCHEMAS LIKE '%s' IN DATABASE IDENTIFIER(?)",
+            sanitizeIdentifierWithWildcardForLikeClause(schema.schemaName()));
+    List<SnowflakeIdentifier> schemas;
+    try {
+      schemas =
+          connectionPool.run(
+              conn ->
+                  queryHarness.query(
+                      conn,
+                      finalQuery,
+                      SCHEMA_RESULT_SET_HANDLER,
+                      new String[] {schema.databaseName()}));
+    } catch (SQLException e) {
+      throw new UncheckedSQLException(e, "Failed to check if schema '%s' exists", schema);
+    } catch (InterruptedException e) {
+      throw new UncheckedInterruptedException(
+          e, "Interrupted while checking if schema '%s' exists", schema);
+    }
+
+    // Filter to handle the edge case of '_' appearing as a wildcard that can't be remapped the way
+    // it can for predicates in SELECT statements.
+    schemas.removeIf(sc -> !schema.schemaName().equalsIgnoreCase(sc.schemaName()));
+    return !schemas.isEmpty();
+  }
+
+  @Override
+  public List<SnowflakeIdentifier> listDatabases() {
+    List<SnowflakeIdentifier> databases;
+    try {
+      databases =
+          connectionPool.run(
+              conn ->
+                  queryHarness.query(
+                      conn,
+                      "SHOW DATABASES IN ACCOUNT",
+                      DATABASE_RESULT_SET_HANDLER,
+                      (String[]) null));
```

Comment:
This is vararg, so you can just leave the param off rather than casting a null.

Reply:
Done